### PR TITLE
support-for-multiple-service-deployment-from-same-repository

### PR DESCRIPTION
**Pull request by Ritik232**

### CODE REVIEW

File: .github/workflows/commonBuild.yml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      version:
+        description: "Version"
+        value: ${{ jobs.build-and-push-image.outputs.version }}
+      cluster_environment:
+        description: "Cluster Environment"
+        value: ${{ jobs.build-and-push-image.outputs.cluster_environment }}
+    secrets:
+      GHCR_TOKEN:
+        required: true
+        description: "secret for GHRC"
```
Comment (ypmittal) [inline .github/workflows/commonBuild.yml:28]: GHRC -> GHCR

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -1,25 +1,36 @@
...
+        description: "secret for Github Container Registry"
       REPOSITORY_OWNER:
         required: true
-        description: 'username of repository owner'
-  push:
-    tags:
-      - v[0-9]+.[0-9]+.[0-9]+-rc[0-9]
-      - v[0-9]+.[0-9]+.[0-9]+
-
+        description: "username of repository owner"
```
Comment (ypmittal) [inline .github/workflows/build.yml:17]: please change to "username of GHCR owner"

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -1,25 +1,36 @@
...
-  workflow_call: 
+  workflow_call:
     outputs:
       version:
-        description: 'Version'
+        description: "Application Version"
         value: ${{ jobs.build-docker.outputs.version }}
       cluster_environment:
-        description: 'Cluster Environment'
+        description: "Cluster Environment-production or testing "
```
Comment (ypmittal) [inline .github/workflows/build.yml:9]: change to "environment where application will be deployed, ex: testing, production, etc."

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -1,25 +1,36 @@
...
-        description: 'username of repository owner'
-  push:
-    tags:
-      - v[0-9]+.[0-9]+.[0-9]+-rc[0-9]
-      - v[0-9]+.[0-9]+.[0-9]+
-
+        description: "username of repository owner"
+    inputs:
+      docker_context_path:
+        description: "docker context path:"
```
Comment (ypmittal) [inline .github/workflows/build.yml:20]: remove ":" at the end

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -1,25 +1,36 @@
...
+        required: false
+        type: string
+        default: "."
+      dockerfile_path:
+        description: "Enter dockerfile path that is will be used to create a image"
```
Comment (ypmittal) [inline .github/workflows/build.yml:25]: change to "path to the Dockerfile"

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -1,25 +1,36 @@
...
+        required: false
+        type: string
+        default: "./Dockerfile"
+      image_name:
+        description: "image name of docker image for current application"
```
Comment (ypmittal) [inline .github/workflows/build.yml:30]: change to "name for the image that will be built (defaults to repository name)"

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -63,10 +77,11 @@ jobs:
...
 
-      - name: Build and push Docker image
+      - name: Build and push Docker image to github container registry
         uses: docker/build-push-action@v3
         with:
-          context: .
-          push: ${{ github.event_name != 'pull_request' }}
+          context: ${{ inputs.docker_context_path }}
+          file: ${{ inputs.dockerfile_path }}
+          push: true
```
Comment (ypmittal) [inline .github/workflows/build.yml:85]: is there a reason we changed `push` behavior?

Comment (Ritik232) [reply .github/workflows/build.yml:85]: I can't find it in the code I pushed.

File: .github/workflows/deploy.yml [hunk 8]
```diff
@@ -4,52 +4,71 @@ on:
   workflow_dispatch:
     inputs:
       repository_name:
-        description: 'Enter repository name:'
+        description: "repository name in which application code is present"
```
Comment (ypmittal) [inline .github/workflows/deploy.yml:7]: change to "name of the service to be deployed (usually it is the same as the repository name)"

File: .github/workflows/deploy.yml [hunk 9]
```diff
@@ -4,52 +4,71 @@ on:
...
         required: true
         type: string
+      image_name:
+        description: "image name of our application's dockerfile"
```
Comment (ypmittal) [inline .github/workflows/deploy.yml:11]: change to "name of the image to deployed (excluding the registry & owner details), ex: atcost-ospos"

File: .github/workflows/deploy.yml [hunk 10]
```diff
@@ -4,52 +4,71 @@ on:
...
+        required: true
+        type: string
+      release_name:
+        description: "application release name"
```
Comment (ypmittal) [inline .github/workflows/deploy.yml:15]: "name of the helm release to be created (should only contain alpha-numeric & hyphen characters)"

File: .github/workflows/deploy.yml [hunk 11]
```diff
@@ -4,52 +4,71 @@ on:
...
-        description: 'Which environment you want to deploy?'
+        description: "Cluster environment to deploy your image"
         required: true
         type: choice
         options:
           - "testing"
           - "production"
       version:
-        description: 'Enter the application version:'
+        description: "application version:"
```
Comment (ypmittal) [inline .github/workflows/deploy.yml:27]: remove ":" at the end

Comment (ypmittal) [reply .github/workflows/deploy.yml:27]: actually change it to "application version (image tag) to be deployed"

File: .github/workflows/deploy.yml [hunk 12]
```diff
@@ -4,52 +4,71 @@ on:
...
         required: true
         type: string
       namespace:
         required: true
-        description: 'Enter the application namespace:'
+        description: "cluseter namespace in which application is going to depolyed"
```
Comment (ypmittal) [inline .github/workflows/deploy.yml:32]: correct cluster spelling, rest is good

File: .github/workflows/deploy.yml [hunk 13]
```diff
@@ -4,52 +4,71 @@ on:
...
         type: string
 
   workflow_call:
     inputs:
       repository_name:
-        description: 'Enter repository name:'
+        description: "repository name in which application code is present"
```
Comment (ypmittal) [inline .github/workflows/deploy.yml:38]: copy changes from workflow_call descriptions

File: .github/workflows/deploy.yml [hunk 14]
```diff
@@ -62,4 +81,4 @@ jobs:
           jenkins-token: "${{ secrets.JENKINS_TOKEN }}"
           user: "${{ secrets.JENKINS_USER }}"
           job-path: "job/kubernetes-deploy-on-nutanix"
-          job-params: '{"SERVICE_NAME" : "${{ inputs.repository_name }}", "CLUSTER_ENVIRONMENT": "${{ inputs.cluster_environment }}" , "VERSION": "${{ inputs.version }}", "NAMESPACE": "${{ inputs.namespace || secrets.NAMESPACE }}" }'
+          job-params: '{"SERVICE_NAME" : "${{ inputs.repository_name }}", "RELEASE_NAME" : "${{ inputs.release_name }}", "IMAGE_NAME" : "${{ inputs.image_name }}", "CLUSTER_ENVIRONMENT": "${{ inputs.cluster_environment }}" , "VERSION": "${{ inputs.version }}", "NAMESPACE": "${{ inputs.namespace || secrets.NAMESPACE }}" }'
```
[the remark this answers is not in the document]
Comment (ypmittal) [reply .github/workflows/deploy.yml:84]: test it as well, i am not sure if this is the correct way to write in multi line

File: .github/workflows/test-basic.yml [hunk 15]
```diff
@@ -0,0 +1,31 @@
+name: Build & publish image
```
Comment (ypmittal) [inline .github/workflows/test-basic.yml:1]: change to "TestCase 1: Basic Repository"

File: .github/workflows/test-monorepo.yml [hunk 16]
```diff
@@ -0,0 +1,45 @@
+name: Build & publish image
```
Comment (ypmittal) [inline .github/workflows/test-monorepo.yml:1]: change to "TestCase 2: Monorepo"

File: .github/workflows/test-one-image-multi-deploy.yml [hunk 17]
```diff
@@ -0,0 +1,34 @@
+name: Build & publish image
```
Comment (ypmittal) [inline .github/workflows/test-one-image-multi-deploy.yml:1]: change to "TestCase 3: Single image multi-deploy"

File: .github/workflows/test-monorepo.yml [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+      contents: read
+      packages: write
+    secrets:
+      GHCR_TOKEN: ${{ secrets.GHCR_TOKEN }}
+      REPOSITORY_OWNER: ${{ secrets.REPOSITORY_OWNER }}  
+    uses: ./.github/workflows/build.yml 
+    with:
+      image_name: ${{ matrix.project }} 
+      docker_context_path: ${{ matrix.project }}
+      dockerfile_path: ./${{ matrix.project }}/Dockerfile
```
Comment (ypmittal) [inline .github/workflows/test-monorepo.yml:24]: bug, create two folders for your projects and copy the root dockerfile in both

File: .github/workflows/test-monorepo.yml [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+       project: [calance-workflows-release-1, calance-workflows-release-2]
+    permissions:
+      contents: read
+      packages: write
+    uses: ./.github/workflows/deploy.yml 
+    with:
+      repository_name: ${{ github.event.repository.name }}
+      version: ${{ needs.build-and-push-dockerimages.outputs.version }}
+      release_name: ${{ matrix.project }}
+      cluster_environment: ${{ needs.build-and-push-dockerimages.outputs.cluster_environment }}
```
Comment (ypmittal) [inline .github/workflows/test-monorepo.yml:40]: bug - need to supply which image to deploy

File: .github/workflows/test-one-image-multi-deploy.yml [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+       project: [calance-workflows-release-1, calance-workflows-release-2]
+    permissions:
+      contents: read
+      packages: write
+    uses: ./.github/workflows/deploy.yml 
+    with:
+      repository_name: ${{ github.event.repository.name }}
+      version: ${{ needs.build-and-push-dockerimages.outputs.version }}
+      release_name: ${{ matrix.project }}
+      cluster_environment: ${{ needs.build-and-push-dockerimages.outputs.cluster_environment }}
```
Comment (ypmittal) [inline .github/workflows/test-one-image-multi-deploy.yml:29]: please supply which image to deploy as well for clarity

File: .github/workflows/deploy.yml [hunk 21]
```diff
@@ -4,52 +4,71 @@ on:
...
 
   workflow_call:
     inputs:
       repository_name:
-        description: 'Enter repository name:'
+        description: "name of the service to be deployed (usually it is the same as the repository name)"
         required: true
         type: string
+      image_name:
+        description: "image name of our application's dockerfile"
```
Comment (ypmittal) [inline .github/workflows/deploy.yml:42]: match all input descriptions with `workflow_dispatch` versions

Comment (ypmittal) [reply .github/workflows/deploy.yml:42]: Do you think we need `workflow_dispatch`?

File: .github/workflows/build.yml (hunk 7 above)
Comment (ypmittal) [reply .github/workflows/build.yml:85]: Please reply to this

Comment (Ritik232) [reply .github/workflows/build.yml:85]: as discussed we are keeping always push behaviour.

**Review by ypmittal — APPROVED** [review]
LGTM

**Pull request merged**